x-pack/auditbeat/tracing/decoder.go

```diff
@@ -0,0 +1,401 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+// +build linux
+
+package tracing
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+	"unsafe"
+)
+
+// This sets a limit on struct decoder's greedy fields. 2048 is not a problem
+// as the kernel tracing subsystem won't let us dump more than that anyway.
+const maxRawCopySize = 2048
+
+// Decoder decodes raw events into an usable type.
+type Decoder interface {
+	// Decode takes a raw message and its metadata and returns a representation
+	// in a decoder-dependent type.
+	Decode(raw []byte, meta Metadata) (interface{}, error)
+}
+
+type mapDecoder []Field
+
+// NewMapDecoder creates a new decoder that will parse raw tracing events
+// into a map[string]interface{}. This decoder will decode all the fields
+// described in the format.
+// The map keys are the field names as given in the format.
+// The map values are fixed-size integers for integer fields:
+// uint8, uint16, uint32, uint64, and for signed fields, their signed counterpart.
+// For string fields, the value is a string.
+// Null string fields will be the null interface.
+func NewMapDecoder(format ProbeFormat) Decoder {
+	fields := make([]Field, 0, len(format.Fields))
+	for _, field := range format.Fields {
+		fields = append(fields, field)
+	}
+	sort.Slice(fields, func(i, j int) bool {
+		return fields[i].Offset < fields[j].Offset
+	})
+	return mapDecoder(fields)
+}
```

Comment:
Nope. It's the first decoder I implemented, but then settled for the struct decoder which is faster and safer because it validates that the types in the structs match the types in the kernel's kprobe format.

Reply:
Ok, lets leave it in case it is useful in the future.